mlir/lib/Quantum/Transforms/ConversionPatterns.cpp

```diff
@@ -585,9 +588,6 @@ template <typename T> struct StateBasedPattern : public OpConversionPattern<T> {
         MLIRContext *ctx = this->getContext();
         TypeConverter *conv = this->getTypeConverter();
 
-        assert(op.getType().template isa<MemRefType>() &&
-               "probs/state must return memref before lowering");
-
```

Comment:
I don't think these are necessary. Unlike before these series of patches, before the result could be either tensor or memref. Here since we are using StateIn which is guaranteed to be a MemRef. With the verifier it is also guaranteed that memrefs and tensors are mutually exclusive.

Reply:
[e002910](https://github.com/PennyLaneAI/catalyst/pull/63/commits/e00291040e0215f142257b26bb2d09bafd95e11e)

Reply:
I guess we could remove them, couldn't the pass be invoked on IR that is not bufferized though?

Actually I think the correct way to handle this is to return a pass failure rather than an assertion 🤔